File: algorithms/simplex.py
import enum
import numpy as np
from dataclasses import dataclass

from model.constraint import Constraint, ConstraintSense
from model.objective import Objective, ObjectiveSense
from algorithms.geometry import EPSILON  # numerical zero for pivot and optimality checks


# --- Data types ---

class OptimizerStatus(enum.Enum):
    OPTIMAL = 'optimal'
    UNBOUNDED = 'unbounded'
    INFEASIBLE = 'infeasible'
    FEASIBLE = 'feasible'
    NONE = 'none'

    def __str__(self) -> str:
        match self:
            case OptimizerStatus.OPTIMAL:
                return 'Linear program has a finite optimum'
            case OptimizerStatus.UNBOUNDED:
                return 'Linear program is unbounded'
            case OptimizerStatus.INFEASIBLE:
                return 'Linear program is infeasible'
            case OptimizerStatus.FEASIBLE:
                return 'Linear program is feasible'
            case OptimizerStatus.NONE:
                return 'Linear program has not been solved yet'
            case _:
                raise ValueError(f'Unrecognized optimization status {self}.')


@dataclass
class Solution:
    point: np.ndarray
    objective_value: float

    def __str__(self) -> str:
        return (f'optimal value {self.objective_value:.3f} '
                f'is attained at ({self.point[0]:.3f}, {self.point[1]:.3f})')


@dataclass
class Result:
    status: OptimizerStatus
    solution: Solution | None

    def __str__(self) -> str:
        if self.status == OptimizerStatus.OPTIMAL and self.solution is not None:
            return str(self.status) + ': ' + str(self.solution)
        return str(self.status)


_TABLEAU_FIXED_COLS = 5   # RHS + x+ + x− + y+ + y−


@dataclass
class SimplexTableau:
    _tableau: np.ndarray
    _num_constraints: int

    @staticmethod
    def _canonical_objective(objective: Objective) -> list[float]:
        """
        Returns the cost row [0, cx+, cx-, cy+, cy-] in canonical (min) form.

        For MIN c^T x: costs are  [ cx,  -cx,  cy, -cy]
        For MAX c^T x: costs are  [-cx,   cx, -cy,  cy]  (negate to minimise)

        Column 0 (RHS) always has cost 0.
        """
        coeff_x, coeff_y = objective.coeff_x, objective.coeff_y
        match objective.sense:
            case ObjectiveSense.MAX:
                return [0., -coeff_x, coeff_x, -coeff_y, coeff_y]
            case ObjectiveSense.MIN:
                return [0., coeff_x, -coeff_x, coeff_y, -coeff_y]

    @staticmethod
    def _canonical_constraint(constraint: Constraint) -> list[list[float]]:
        """
        Returns one or two canonical rows for the constraint in the form:
            [rhs, cx+, cx-, cy+, cy-]

        LE:  1 row  (kept as-is)
        GE:  1 row  (negated -> equivalent LE)
        EQ:  2 rows (LE row + GE row)
        """
        rhs, coeff_x, coeff_y = constraint.rhs, constraint.coeff_x, constraint.coeff_y
        match constraint.sense:
            case ConstraintSense.LE:
                return [[rhs, coeff_x, -coeff_x, coeff_y, -coeff_y]]
            case ConstraintSense.GE:
                return [[-rhs, -coeff_x, coeff_x, -coeff_y, coeff_y]]
            case ConstraintSense.EQ:
                return [[rhs, coeff_x, -coeff_x, coeff_y, -coeff_y],
                        [-rhs, -coeff_x, coeff_x, -coeff_y, coeff_y]]

    @classmethod
    def canonical_from(cls, objective: Objective, constraints: list[Constraint]) -> SimplexTableau:
        """
        Build the initial simplex tableau in canonical form:

            b | A | I
            ---------
            0 | c | 0

        where the columns are [RHS | cx+ | cx- | cy+ | cy- | slacks].
        """
        objective_row = np.array(SimplexTableau._canonical_objective(objective))

        if len(constraints) == 0:
            return cls(_tableau=np.array([objective_row]), _num_constraints=0)

        constraint_rows: list[list[float]] = []
        for constraint in constraints:
            constraint_rows.extend(cls._canonical_constraint(constraint))
        num_constraints = len(constraint_rows)

        return cls(
            _tableau=np.vstack([
                np.hstack([constraint_rows, np.eye(num_constraints)]),
                np.hstack([objective_row, np.zeros(num_constraints)]),
            ]),
            _num_constraints=num_constraints,
        )

    def get_initial_basis(self) -> list[int]:
        return list(range(_TABLEAU_FIXED_COLS, self.num_cols()))

    def is_rhs_all_non_negative(self) -> bool:
        return bool(np.all(self._tableau[:self._num_constraints, 0] >= -EPSILON))

    def get_index_smallest_cost(self) -> int:
        return int(self._tableau[-1, 1:].argmin()) + 1

    def get_objective(self) -> np.ndarray:
        return self._tableau[-1]

    def get_objective_value(self) -> float:
        return self._tableau[-1, 0]

    def get_rhs(self, row: int) -> float:
        return self._tableau[row, 0]

    def num_rows(self) -> int:
        return self._tableau.shape[0]

    def num_cols(self) -> int:
        return self._tableau.shape[1]

    def min_ratio(self, col: int) -> int | None:
        """
        Bland's rule / minimum-ratio test. Returns the leaving row index
        or None if no positive entry exists (unbounded direction).
        """
        ratios = [
            (self.get_rhs(row) / self._tableau[row, col], row)
            for row in range(self._num_constraints)
            if self._tableau[row, col] > EPSILON
        ]
        return min(ratios, default=(None, None))[1]

    def pivot_row_operation(self, curr_row: int, pivot_row: int, pivot_col: int):
        self._tableau[curr_row] -= self._tableau[curr_row, pivot_col] * self._tableau[pivot_row]

    def pivot(self, row: int, col: int) -> None:
        """Pivot tableau on element (row, col) in-place"""
        self._tableau[row] /= self._tableau[row, col]
        for i in range(self._tableau.shape[0]):
            if i != row:
                self.pivot_row_operation(curr_row=i, pivot_row=row, pivot_col=col)

    def phase_1_add_objective(self, row: np.ndarray):
        self._tableau = np.vstack([self._tableau, row])

    def phase_1_add_artificial_variable(self, row: int):
        self._tableau[row] *= -1

        artificial_column = np.zeros((self._tableau.shape[0], 1))
        artificial_column[row,  0] = 1.0
        artificial_column[-1, 0] = 1.0

        self._tableau = np.hstack([self._tableau, artificial_column])

    def phase_1_drop_objective(self) -> None:
        self._tableau = self._tableau[:-1]

    def phase_1_drop_row(self, row: int) -> None:
        self._tableau = np.delete(self._tableau, row, axis=0)
        self._num_constraints -= 1

    def phase_1_remove_artificial_variables(self):
        self._tableau = self._tableau[:, :_TABLEAU_FIXED_COLS]


@dataclass
class SimplexSolver:
    _status: OptimizerStatus
    _sense: ObjectiveSense
    _tableau: SimplexTableau
    _basis: list[int]
    _artificial_variables: list[int]

    def _iterate(self) -> OptimizerStatus:
        """
        Drive simplex pivots on current tableau (last row = objective).

        Mutates _tableau and _basis in-place.

        Returns:
            Status.OPTIMAL: if all reduced costs >= 0; current BFS is optimal.
            Satus.UNBOUNDED: if an entering column has no positive constraint entries;
                the objective is unbounded.
        """
        while True:
            enter_col = self._tableau.get_index_smallest_cost()
            if self._tableau.get_objective()[enter_col] >= -EPSILON:
                return OptimizerStatus.OPTIMAL

            leave_row = self._tableau.min_ratio(enter_col)
            if leave_row is None:
                return OptimizerStatus.UNBOUNDED

            self._tableau.pivot(leave_row, enter_col)
            self._basis[leave_row] = enter_col

    def _phase_1_setup(self) -> None:
        """
        Append a phase-1 objective row (minimise sum of artificials),
        then for every constraint row whose RHS is negative:
        negate the row and add a fresh artificial variable column.
        """
        num_cols = self._tableau.num_cols()

        self._tableau.phase_1_add_objective(np.zeros(num_cols))

        self._artificial_variables = []
        for row in range(self._tableau._num_constraints):  # todo: fix encapsulation violation (self._tableau._num_constraints)
            if self._tableau.get_rhs(row) < -EPSILON:
                self._tableau.phase_1_add_artificial_variable(row)
                self._tableau.pivot_row_operation(curr_row=-1, pivot_row=row, pivot_col=num_cols)
                self._artificial_variables.append(num_cols)
                self._basis[row] = num_cols
                num_cols += 1

    def _phase_1_teardown(self) -> None:
        """
        Drop the phase-1 objective row, pivot out any artificial that is still
        basic at zero (degenerate BFS), then remove all artificial columns and
        re-index the basis.
        """
        # todo: fix encapsulation violations (self._tableau._tableau)

        self._tableau.phase_1_drop_objective()

        art_set = set(self._artificial_variables)
        first_artificial = min(self._artificial_variables)

        for row in range(len(self._basis) - 1, -1, -1):
            bv = self._basis[row]
            if bv not in art_set:
                continue
            swapped = False
            for j in range(first_artificial): # only non-artificial cols
                if abs(self._tableau._tableau[row, j]) > EPSILON:
                    self._tableau.pivot(row, j)
                    self._basis[row] = j
                    swapped = True
                    break
            if not swapped:
                # Row is linearly dependent — drop it.
                self._tableau.phase_1_drop_row(row)
                self._basis.pop(row)

        # Remove artificial columns and re-index basis.
        total_cols = self._tableau.num_cols()
        keep = [j for j in range(total_cols) if j not in art_set]
        self._tableau._tableau = self._tableau._tableau[:, keep]
        col_map = {old: new for new, old in enumerate(keep)}
        self._basis = [col_map[bv] for bv in self._basis]

    def _phase_1(self) -> None:
        self._phase_1_setup()

        self._iterate()

        if abs(self._tableau.get_objective_value()) > EPSILON:
            self._status = OptimizerStatus.INFEASIBLE
            return

        self._phase_1_teardown()
        self._status = OptimizerStatus.FEASIBLE

    @classmethod
    def solve(cls, objective: Objective, constraints: list[Constraint]) -> SimplexSolver:
        tableau = SimplexTableau.canonical_from(objective=objective, constraints=constraints)
        solver = SimplexSolver(
            _status=OptimizerStatus.NONE,
            _sense=objective.sense,
            _tableau=tableau,
            _basis=tableau.get_initial_basis(),
            _artificial_variables=[],
        )

        if solver._tableau.is_rhs_all_non_negative():
            solver._status = OptimizerStatus.FEASIBLE
        else:
            solver._phase_1()

        if solver._status is OptimizerStatus.FEASIBLE:
            solver._status = solver._iterate()

        return solver

    def _get_solution(self) -> Solution | None:
        if self._status != OptimizerStatus.OPTIMAL:
            return None

        basic_values: dict[int, float] = {
            basic_variable: self._tableau.get_rhs(row)
            for row, basic_variable in enumerate(self._basis)
        }
        x = basic_values.get(1, 0.0) - basic_values.get(2, 0.0)
        y = basic_values.get(3, 0.0) - basic_values.get(4, 0.0)
        point = np.array([x, y])

        objective_value = self._tableau.get_objective_value()
        if self._sense is ObjectiveSense.MIN:
            objective_value = -objective_value

        return Solution(point=point, objective_value=objective_value)

    def get_result(self) -> Result:
        return Result(status=self._status, solution=self._get_solution())

File: lp-visualizer-clean/src/simplex.ts
import { ConstraintSense, EPSILON, ObjectiveSense, OptimizerStatus } from './types';
import type {
  Constraint,
  LPResult,
  Objective,
  Solution,
} from './types';

// ── Matrix utilities ──────────────────────────────────────────────────────────

type Matrix = number[][];

function rowScale(m: Matrix, row: number, factor: number): void {
  for (let j = 0; j < m[row].length; j++) m[row][j] *= factor;
}

function rowAdd(m: Matrix, target: number, source: number, factor: number): void {
  for (let j = 0; j < m[target].length; j++) m[target][j] += factor * m[source][j];
}

// ── SimplexTableau ────────────────────────────────────────────────────────────

const FIXED_COLS = 5; // RHS + x+ x- + y+ y-

class SimplexTableau {
  tableau: Matrix;
  numConstraints: number;

  constructor(tableau: Matrix, numConstraints: number) {
    this.tableau = tableau;
    this.numConstraints = numConstraints;
  }

  private static canonicalObjectiveRow(obj: Objective): number[] {
    const { coeffX: cx, coeffY: cy, sense } = obj;
    return sense === ObjectiveSense.MAX
      ? [0, -cx,  cx, -cy,  cy]
      : [0,  cx, -cx,  cy, -cy];
  }

  private static canonicalConstraintRows(c: Constraint): number[][] {
    const { coeffX: cx, coeffY: cy, rhs } = c;
    const le = [ rhs,  cx, -cx,  cy, -cy];
    const ge = [-rhs, -cx,  cx, -cy,  cy];
    switch (c.sense) {
      case ConstraintSense.LE: return [le];
      case ConstraintSense.GE: return [ge];
      case ConstraintSense.EQ: return [le, ge];
      default:                 return [le];
    }
  }

  static canonicalFrom(obj: Objective, constraints: Constraint[]): SimplexTableau {
    const objRow = SimplexTableau.canonicalObjectiveRow(obj);

    if (constraints.length === 0) {
      return new SimplexTableau([[...objRow]], 0);
    }

    const cRows: number[][] = [];
    for (const c of constraints) cRows.push(...SimplexTableau.canonicalConstraintRows(c));

    const n = cRows.length;
    const block: Matrix = cRows.map((row, i) => {
      const slacks = new Array<number>(n).fill(0);
      slacks[i] = 1;
      return [...row, ...slacks];
    });

    const fullObjRow = [...objRow, ...new Array<number>(n).fill(0)];
    return new SimplexTableau([...block, fullObjRow], n);
  }

  getInitialBasis(): number[] {
    return Array.from({ length: this.numConstraints }, (_, i) => FIXED_COLS + i);
  }

  isRhsAllNonNegative(): boolean {
    for (let i = 0; i < this.numConstraints; i++) {
      if (this.tableau[i][0] < -EPSILON) return false;
    }
    return true;
  }

  getIndexSmallestCost(): number {
    const obj = this.tableau[this.tableau.length - 1];
    let minIdx = 1;
    for (let j = 2; j < obj.length; j++) {
      if (obj[j] < obj[minIdx]) minIdx = j;
    }
    return minIdx;
  }

  getObjectiveRow(): number[] { return this.tableau[this.tableau.length - 1]; }
  getObjectiveValue(): number { return this.tableau[this.tableau.length - 1][0]; }
  getRhs(row: number): number { return this.tableau[row][0]; }
  numRows(): number           { return this.tableau.length; }
  numCols(): number           { return this.tableau[0]?.length ?? 0; }

  minRatio(col: number): number | null {
    let minR = Infinity;
    let minRow: number | null = null;
    for (let row = 0; row < this.numConstraints; row++) {
      const entry = this.tableau[row][col];
      if (entry > EPSILON) {
        const ratio = this.getRhs(row) / entry;
        if (ratio < minR) { minR = ratio; minRow = row; }
      }
    }
    return minRow;
  }

  pivotRowOp(currRow: number, pivotRow: number, pivotCol: number): void {
    rowAdd(this.tableau, currRow, pivotRow, -this.tableau[currRow][pivotCol]);
  }

  pivot(row: number, col: number): void {
    rowScale(this.tableau, row, 1 / this.tableau[row][col]);
    for (let i = 0; i < this.tableau.length; i++) {
      if (i !== row) this.pivotRowOp(i, row, col);
    }
  }

  phase1AddObjectiveRow(row: number[]): void {
    this.tableau.push([...row]);
  }

  phase1AddArtificialVar(row: number): void {
    this.tableau[row] = this.tableau[row].map(v => -v);
    const lastRow = this.tableau.length - 1;
    for (let i = 0; i < this.tableau.length; i++) {
      this.tableau[i].push(i === row || i === lastRow ? 1 : 0);
    }
  }

  phase1DropObjectiveRow(): void { this.tableau.pop(); }

  phase1DropRow(row: number): void {
    this.tableau.splice(row, 1);
    this.numConstraints--;
  }

  phase1Teardown(artificialVars: number[], basis: number[]): number[] {
    this.phase1DropObjectiveRow();

    const artSet  = new Set(artificialVars);
    const firstArt = Math.min(...artificialVars);

    for (let row = this.numConstraints - 1; row >= 0; row--) {
      if (!artSet.has(basis[row])) continue;

      let swapped = false;
      for (let j = 0; j < firstArt; j++) {
        if (Math.abs(this.tableau[row][j]) > EPSILON) {
          this.pivot(row, j);
          basis[row] = j;
          swapped = true;
          break;
        }
      }

      if (!swapped) {
        this.phase1DropRow(row);
        basis.splice(row, 1);
      }
    }

    const totalCols = this.numCols();
    const keep = Array.from({ length: totalCols }, (_, j) => j).filter(j => !artSet.has(j));
    this.tableau = this.tableau.map(row => keep.map(j => row[j]));

    const colMap = new Map(keep.map((oldJ, newJ) => [oldJ, newJ]));
    return basis.map(bv => colMap.get(bv)!);
  }
}

// ── SimplexSolver ─────────────────────────────────────────────────────────────

class SimplexSolver {
  status: OptimizerStatus;
  private readonly sense: ObjectiveSense;
  private readonly tableau: SimplexTableau;
  private basis: number[];
  private artificialVars: number[] = [];

  constructor(
    status: OptimizerStatus,
    sense: ObjectiveSense,
    tableau: SimplexTableau,
    basis: number[],
  ) {
    this.status  = status;
    this.sense   = sense;
    this.tableau = tableau;
    this.basis   = basis;
  }

  private iterate(): OptimizerStatus {
    while (true) {
      const enterCol = this.tableau.getIndexSmallestCost();
      if (this.tableau.getObjectiveRow()[enterCol] >= -EPSILON) {
        return OptimizerStatus.OPTIMAL;
      }
      const leaveRow = this.tableau.minRatio(enterCol);
      if (leaveRow === null) return OptimizerStatus.UNBOUNDED;
      this.tableau.pivot(leaveRow, enterCol);
      this.basis[leaveRow] = enterCol;
    }
  }

  private phase1Setup(): void {
    let numCols = this.tableau.numCols();
    this.tableau.phase1AddObjectiveRow(new Array<number>(numCols).fill(0));

    for (let row = 0; row < this.tableau.numConstraints; row++) {
      if (this.tableau.getRhs(row) < -EPSILON) {
        this.tableau.phase1AddArtificialVar(row);
        this.tableau.pivotRowOp(this.tableau.numRows() - 1, row, numCols);
        this.artificialVars.push(numCols);
        this.basis[row] = numCols;
        numCols++;
      }
    }
  }

  private phase1(): void {
    this.phase1Setup();
    this.iterate();

    if (Math.abs(this.tableau.getObjectiveValue()) > EPSILON) {
      this.status = OptimizerStatus.INFEASIBLE;
      return;
    }

    this.basis  = this.tableau.phase1Teardown(this.artificialVars, this.basis);
    this.status = OptimizerStatus.FEASIBLE;
  }

  private getSolution(): Solution | null {
    if (this.status !== OptimizerStatus.OPTIMAL) return null;

    const vals = new Map<number, number>();
    for (let row = 0; row < this.basis.length; row++) {
      vals.set(this.basis[row], this.tableau.getRhs(row));
    }

    const x = (vals.get(1) ?? 0) - (vals.get(2) ?? 0);
    const y = (vals.get(3) ?? 0) - (vals.get(4) ?? 0);

    let objValue = this.tableau.getObjectiveValue();
    if (this.sense === ObjectiveSense.MIN) objValue = -objValue;

    return { point: [x, y], objectiveValue: objValue };
  }

  getResult(): LPResult {
    return { status: this.status, solution: this.getSolution() };
  }

  static solve(obj: Objective, constraints: Constraint[]): LPResult {
    const tableau = SimplexTableau.canonicalFrom(obj, constraints);
    const solver  = new SimplexSolver(
      OptimizerStatus.NONE,
      obj.sense,
      tableau,
      tableau.getInitialBasis(),
    );

    if (solver.tableau.isRhsAllNonNegative()) {
      solver.status = OptimizerStatus.FEASIBLE;
    } else {
      solver.phase1();
    }

    if (solver.status === OptimizerStatus.FEASIBLE) {
      solver.status = solver.iterate();
    }

    return solver.getResult();
  }
}

// ── Public API ────────────────────────────────────────────────────────────────

export function solveLp(obj: Objective, constraints: Constraint[]): LPResult {
  return SimplexSolver.solve(obj, constraints);
}
